### ISSUE TITLE
How to distribute/push enterprise app to enrolled iOS device?

### ISSUE THREAD

**Issue by the-c0d3r** [micromdm/micromdm#512]
### What version of `micromdm` are you using?
1.3.2

I'm using MicroMDM to manage an enterprise iOS devices, and I'm trying to install enterprise applications. And it's not using DEP.

I was trying to push an enterprise app developed in-house for enterprise devices. I Read through the documentation, I can only see the Wiki page [distributing packages with InstallApplication](https://github.com/micromdm/micromdm/wiki/distributing-packages-with-InstallApplication). This seems to be intendend for macOS packages instead of iOS. And this [API Docs](https://github.com/micromdm/micromdm/wiki/API-Docs) on calling the API endpoint. 

I have the `Manifest.plist` generated by Xcode, and it's IPA file. Can someone guide me on how to install the ipa to the enrolled device remotely? I have the Enterprise Developer License, and I'm new on iOS development. 

According to the wiki page I linked earlier, I need to host manifest file on the mdm server's web root. But I cannot find the web root under the MDM folder. Can anyone tell me which directory is the web root and how to host file to it?

**Comment by jessepeterson** [comment]
Are you running MicroMDM now? The `-filerepo` switch under the `serve` will start a simple little file-serving web server on MicroMDM's along with the MDM service. Also note you don't have to use MicroMDM's web server for this. It can be hosted on any website if you like — the MDM command tells it where to access to the manifest. And finally note that the new `InstallEnterpriseApplication` command actually _embeds_ the Manifest in the MDM command directly.

**Comment by the-c0d3r** [comment]
Hey @jessepeterson thanks for your comment. I tried the `-filerepo` argument. And tried accessing the path from browser. It is 404. But now I'm serving it with python and handling the redirections with nginx. 

After launching the `InstallEnterpriseApplication` api file, webhook responded the following. I have verified the manifest.plist's urls are all accessible.

```
{
    "topic": "mdm.Connect", 
    "event_id": "22aa4905-d489-4759-bbdc-87ad55529d0e",
    "created_at": "2018-09-25T06:54:49.226921976Z", 
    "acknowledge_event": {
        "udid": "xxxxxxx", 
        "status": "Error", 
        "url_parms": None, 
        "command_uuid": "064f9304-d76d-4078-bf97-fcf634c41bc0", 
        "raw_payload": "xxxx"
    }
}
```

Below is the raw_payload decoded

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com¯DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>CommandUUID</key>
        <string>064f9304-d76d-4078-bf97-fcf634c41bc0</string>
        <key>ErrorChain</key>
        <array>
            <dict>
                <key>ErrorCode</key>
                <integer>12021</integer>
                <key>ErrorDomain</key>
                <string>MCMDMErrorDomain</string>
                <key>LocalizedDescription</key>
                <string>âInstallEnterpri³eApplicationâ is not a valid request type.</string>
                <key>USEnglishDescription</key>
                <string>âInstallEnterpriseApplicationâ is not a valid request type.</string>
            </¼ict>
        </array>
        <key>Status</key>
        <string>Error</string>
        <key>UDID</key>
        <string>xxxxx</string>
    </dict>
</plist>
```

I tried multiple times, it still give the same result. May I know if I did anything wrong?

**Comment by jessepeterson** [comment]
Not sure why `-filerepo` didn't work for you. I'm not sure what you put under the repository and which path you tried to download, etc.

But, the good news: you're getting commands successfully to the device and successfully getting responses. MDM enrollment and communication side is good! I'd consider that the 'hard part' done. You just need to troubleshoot the specific command you're sending.

What version device are you sending to? It might be too early to support IEP (is a new command). Try the InstallApplication. Do _any_ commands work (e.g. deviceinformation, securityinfo, etc.).

**Comment by the-c0d3r** [comment]
Hi, I used the `-filerepo` method with the following args `-filerepo ipa/` then tried to access the web root's ipa directory. It doesn't list the dir, but instead 404, as well as accessing file with file name. 

My device is 11.4.1. Does micromdm only support certain versions? 
I did the `InstallApplication`. It shows a confirmation prompt on the device and I can install it with that. But my app is supposed to be pushed silently, as it is a compulsory enterprise app that user must install. 

My ipa was generated by exporting the app as "iOS App Store" in "select distribution mode". I thought it was the reason InstallEnterprise does not work. I exported the app again as "Enterprise" in distribution mode. It also gives the same output. Is there any other way to install without asking the user to "confirm"?
Apart from that, every other command works.

**Comment by jessepeterson** [comment]
Sounds like you got `InstallApplication` working, that's great! To my knowledge the only way to have a silently installed iOS app is to have the device either Supervised or enrolled via DEP. Is that how you're enrolling devices? MicroMDM should support all 'modern' iOS versions (mostly limited by TLS ciphers). It's just that the `InstallEnterpriseApplication` is only supported on 11.4 maybe? Or maybe iOS 12?

**Comment by jessepeterson** [comment]
Here's a little more (high-level) info from our SimpleMDM pals: https://simplemdm.com/2016/08/09/install-apps-silently/

**Comment by the-c0d3r** [comment]
Hey, thanks for your suggestion. I managed to get the `-filerepo` working. It was my bad. I didn't catch that `-filerepo`'s web root is in repo, not the dir specified. And it doesn't list the directory index when called with `https://server/repo`. Only when I access with exact file names, then it doesn't display 404 anymore. 

As for my devices, I'm not enrolling them in DEP, or Supervised method. Our policy is more like BYOD than company issue. By the way, I tested the `InstallEnterpriseApplication` on both 11.4 and 12. It doesn't seem to work on both. Maybe like you said it only work for DEP or Supervised mode. Thanks for answering and for this opensource project.
